### PR TITLE
Add `IntoIterator` instances for Optional and Result

### PR DESCRIPTION
**Pull request by fosskers**
While nearly trivial, these are still quite reasonable to have, and are added in the spirit of "if a lawful instance is possible, then write it".

Let me know if you'd like tests written.

### CODE REVIEW

File: library/iterator.lisp [hunk 1]
```diff
@@ -147,6 +147,22 @@ Behavior is undefined if the iterator is advanced after a destructive modificati
   (define-instance (IntoIterator (Iterator :elt) :elt)
     (define into-iter id))
 
+  (define-instance (IntoIterator (Optional :elt) :elt)
+    (define (into-iter opt)
+      "Yield the single value of this OPT, or nothing."
+      (match opt
+        ((Some a) (let ((cell (cell:new (make-list a))))
+                    (%Iterator (fn () (cell:pop! cell)))))
```
Comment (fosskers) [inline library/iterator.lisp:155]: I borrowed the `cell` technique from the `List` instance of this typeclass.

**Comment by eliaslfox** [comment]
I don't think this needs to allocate a list. It should be possible to store a boolean flag in a cell instead.

**Comment by fosskers** [comment]
Let me try.

**Comment by gefjon** [comment]
It should probably look like
```
(define-instance (IntoIterator (Optional :elt) :elt)
  (define (into-iter opt)
    (match opt
      ((None) empty)
      ((Some elt)
       (let done? = (cell:new False))
       (new (fn ()
              (if (cell:read done?)
                  None
                  (Some elt))))))))
```

**Review by eliaslfox — APPROVED** [review]
LGTM

**Comment by fosskers** [comment]
Updated. It's what @gefjon suggested just in a different shape.

**Pull request merged**